src/timer.tsx

```diff
@@ -0,0 +1,438 @@
+}
+
+const shouldRenderUnit = (
+    unit: TimeUnit,
+    labelType?: Label,
+    minTimeUnit?: TimeUnit,
+    maxTimeUnit?: TimeUnit
+) => {
+    if (!minTimeUnit && !maxTimeUnit) {
+        minTimeUnit = 'seconds';
+        maxTimeUnit = 'hours';
+    }
+    minTimeUnit = minTimeUnit ?? 'seconds';
+    maxTimeUnit = maxTimeUnit ?? 'days';
+
+    // If label is "none", days shouldn't be displayed
+    minTimeUnit = labelType === 'none' && minTimeUnit === 'days' ? 'hours' : minTimeUnit;
+    maxTimeUnit = labelType === 'none' && maxTimeUnit === 'days' ? 'hours' : maxTimeUnit;
+
+    const unitsOrder: Array<TimeUnit> = ['seconds', 'minutes', 'hours', 'days'];
+
+    const minValue = unitsOrder.indexOf(minTimeUnit);
+    // If max < min, we display only the min unit
+    const maxValue = Math.max(unitsOrder.indexOf(maxTimeUnit), minValue);
+    const unitValue = unitsOrder.indexOf(unit);
+
+    return minValue <= unitValue && unitValue <= maxValue;
+};
+
+const getFilteredTimerValue = (
+    timestamp: Timestamp,
+    labelType?: Label,
+    minTimeUnit?: TimeUnit,
+    maxTimeUnit?: TimeUnit
+) => {
+    return (
+        [
+            {unit: 'days', value: timestamp.days},
+            {unit: 'hours', value: timestamp.hours},
+            {unit: 'minutes', value: timestamp.minutes},
+            {unit: 'seconds', value: timestamp.seconds},
+        ] as Array<{unit: TimeUnit; value: number}>
+    ).filter((item) => shouldRenderUnit(item.unit, labelType, minTimeUnit, maxTimeUnit));
+};
+
+const getRemainingTime = (endTimestamp: Date | number) => {
+    // Always return 0 ms remaining for screenshot tests to avoid unstable values caused by delays in browser
+    if (isRunningAcceptanceTest()) {
+        return {
+            days: 0,
+            hours: 0,
+            minutes: 0,
+            seconds: 0,
+        };
+    }
+    const remainingTime = Math.max(
+        0,
+        (typeof endTimestamp === 'object' ? endTimestamp : new Date(endTimestamp)).valueOf() - Date.now()
+    );
+
+    return {
+        days: Math.floor(remainingTime / DAY_IN_MS),
+        hours: Math.floor(remainingTime / HOUR_IN_MS) % 24,
+        minutes: Math.floor(remainingTime / MINUTE_IN_MS) % 60,
+        seconds: Math.floor(remainingTime / SECOND_IN_MS) % 60,
+    };
+};
+
+const useTimerState = ({
+    endTimestamp,
+    labelType,
+    minTimeUnit,
+    maxTimeUnit,
+    onProgress,
+}: {
+    endTimestamp: Date | number;
+    labelType?: Label;
+    minTimeUnit?: TimeUnit;
+    maxTimeUnit?: TimeUnit;
+    onProgress?: (value: Timestamp) => void;
+}) => {
+    const remainingTime = getRemainingTime(endTimestamp);
+    const [currentDays, setCurrentDays] = React.useState(remainingTime.days);
+    const [currentHours, setCurrentHours] = React.useState(remainingTime.hours);
+    const [currentMinutes, setCurrentMinutes] = React.useState(remainingTime.minutes);
+    const [currentSeconds, setCurrentSeconds] = React.useState(remainingTime.seconds);
```

Comment:
why do we need all these states? I think a single state with the remaining time should be enough, the other states can be derived from it at render time

Reply:
Updated